### PR TITLE
fix: nano-contract history pagination and list order

### PR DESCRIPTION
**Pull request by alexruzenhack**
### Acceptance Criteria
- It should fix nano contract transaction loading for newer transactions
- Add 'Load More' in the footer of transactions list, which triggers the request for older transactions when pressed
  - Hides the 'Load More' button when the last transaction of the list is the 'initialize'

#### Loading newer transactions

https://github.com/user-attachments/assets/781e6759-6db2-4e81-b3bc-c20641c097ed

#### Loading older transactions

https://github.com/user-attachments/assets/68f99f4e-b8be-4cf2-a02f-63f1268bfa18

The video above shows a list of transactions initialized with the newest transaction only. When the button 'Load More' is pressed a list of older transactions is loaded. The button 'Load More' didn't show after load because the list has reached its end.

---

Related issues:
- [ ] https://github.com/HathorNetwork/hathor-wallet-mobile/issues/514

---

### Security Checklist
- [x] Make sure you do not include new dependencies in the project unless strictly necessary and do not include dev-dependencies as production ones. More dependencies increase the possibility of one of them being hijacked and affecting us.

### CODE REVIEW

File: src/sagas/nanoContract.js [hunk 1]
```diff
@@ -172,21 +171,37 @@ export function* registerNanoContract({ payload }) {
  * @param {string} after Transaction hash to start to get items
  * @param {Object} wallet Wallet instance from redux state
  *
+ * @description
+ * We use `after` because we look at time from present to future, therefore `after` means
```
Comment (andreabadesso) [inline src/sagas/nanoContract.js:175]: I don't get it... We are actually fetching transactions "before" a given transaction as the user "infinite scrolls" deeper, shouldn't we name the variable `before`?

File: src/sagas/nanoContract.js [hunk 2]
```diff
@@ -172,21 +171,37 @@ export function* registerNanoContract({ payload }) {
...
   if (!success) {
     throw new Error('Failed to fetch nano contract history');
   }
-
-  const history = [];
-  for (const rawTx of rawHistory) {
+  // We are interested to produce a list of transactions in descending order
+  // because we want users to see newest txs first.
+  const historyReversed = new Array(rawHistory.length)
+  for (let idx = 0; idx < rawHistory.length; idx += 1) {
```
Comment (andreabadesso) [inline src/sagas/nanoContract.js:203]: suggestion(non-blocking): You could refactor this to use `Promise.all` so those multiple async requests would run concurrently

File: src/sagas/nanoContract.js [hunk 3]
```diff
@@ -213,15 +228,18 @@ export async function fetchHistory(ncId, count, after, wallet) {
...
-    history.push(tx);
+    historyReversed[rawHistory.length - idx - 1] = tx;
   }
 
   let next = after;
-  if (history && history.length > 0) {
-    next = history[history.length - 1].txId;
+  if (historyReversed && historyReversed.length > 0) {
+    // We should get the first item of the list because we are looking
+    // from present -> past, and we want to paginate from preset toward future.
```
Comment (andreabadesso) [inline src/sagas/nanoContract.js:237]: I'm confused, we're actually paginating towards the past, right?

You scroll to see earlier transactions, i.e. the latest transaction is on top

**Review by andreabadesso — CHANGES_REQUESTED** [review]

File: src/components/NanoContract/NanoContractDetails.js [hunk 4]
```diff
@@ -86,6 +86,8 @@ export const NanoContractDetails = ({ nc }) => {
        * it will be set during the load.
        */
       dispatch(nanoContractHistoryRequest({ ncId: nc.ncId, after: null }));
+    } else {
+      dispatch(nanoContractHistoryRequest({ ncId: nc.ncId, after }))
```
Comment (andreabadesso) [inline src/components/NanoContract/NanoContractDetails.js:90]: You should add `after` as a dependency for the `useEffect`

**Review by andreabadesso — CHANGES_REQUESTED** [review]

File: src/sagas/nanoContract.js (hunk 1 above)
Comment (alexruzenhack) [reply src/sagas/nanoContract.js:175]: Sorry for the confusion, let me try again:
When we call `fetchHistory` with `after` param this denotes that we want new transactions _after_ a certain tx, and we are looking from the present moment toward the future. This is a common knowledge of how to think in time, _after_ means the future.

{txs that we have (past)} **{after (the last tx that we have)}** {txs we are interested in (future)}

However the method `getNanoContractHistory` uses the following signature:
`getNanoContractHistory(ncId, count, after, before)`.

Here, `after` denotes all the transactions from a certain tx towards the **past**. I believe it has something to do with the way we walk the dag. However, this way to frame time is not like the way we usually do, and it can makes things really odd in my opinion, and that's why I'm swapping the words `before` to `after`. After all, we are interested in the future when we ask a new chunk of transactions.

File: src/sagas/nanoContract.js (hunk 2 above)
Comment (alexruzenhack) [reply src/sagas/nanoContract.js:203]: But if we do this we can't use guarantee the order of execution, right? I need it ordered.

File: src/sagas/nanoContract.js (hunk 3 above)
Comment (alexruzenhack) [reply src/sagas/nanoContract.js:237]: This is confusing, indeed. I believe I can remove this comment. We just need to assign the newest tx hash to `next`, that's it.

File: src/components/NanoContract/NanoContractDetails.js (hunk 4 above)
Comment (alexruzenhack) [reply src/components/NanoContract/NanoContractDetails.js:90]: If we do this, the `nanoContractHistoryRequest` will be dispatched twice. The other side effect would be the dispatch of `nanoContractHistoryRequest` until we react the newest tx, however it would happen without a user action, which is not good for usability. `txHistory` already provokes a new rendering if it change, I think.

File: src/sagas/nanoContract.js (hunk 1 above)
Comment (pedroferreira1) [reply src/sagas/nanoContract.js:175]: The words `after` and `before` doesn't have any relation with future and past in this API. I think it's confusing to swap the words, we should use the same words as it's used in the API. `after` is the pagination parameter to get the history after a specific tx, and `before` the same. The API returns the transactions in the descending order, i.e., from the newest to the oldest, as it's expected in the screen.

File: src/sagas/nanoContract.js (hunk 2 above)
Comment (andreabadesso) [reply src/sagas/nanoContract.js:204]: You could pass the index to the promise and return it like [index, response], then use it to assembly the array

But it's non-blocking at all, if you think it's not worth the effort, don't do it

Comment (alexruzenhack) [reply src/sagas/nanoContract.js:212]: Ok. I have opened an issue for it: https://github.com/HathorNetwork/hathor-wallet-mobile/issues/514

File: src/sagas/nanoContract.js (hunk 3 above)
Comment (alexruzenhack) [reply src/sagas/nanoContract.js]: I have refactored it, can you take a new look?

File: src/components/NanoContract/NanoContractDetails.js (hunk 4 above)
Comment (alexruzenhack) [reply src/components/NanoContract/NanoContractDetails.js]: I have refactored it, can you take a new look?

File: src/sagas/nanoContract.js (hunk 1 above)
Comment (alexruzenhack) [reply src/sagas/nanoContract.js]: Ok. I just refactored it. Can you both take a new look at it?

File: src/sagas/nanoContract.js [hunk 5]
```diff
@@ -93,7 +93,10 @@ export function* registerNanoContract({ payload }) {
     return;
   }
 
-  const isAddressMine = yield call(wallet.isAddressMine.bind(wallet), address);
+  // XXX: Make sure to disable the Wallet Service feature if you are testing
+  // the nano-testnet network because Wallet Service Wallet doesn't implement
+  // `isAddressMine`.
```
Comment (andreabadesso) [inline src/sagas/nanoContract.js:98]: I think this should be done automatically as we might eventually want to control the nano-testnet in Unleash

Is there an issue for this? I don't think it should be done in this PR

Comment (andreabadesso) [reply src/sagas/nanoContract.js:98]: In this PR, you could just set it to false if the user is in the wallet-service, so it doesn't crash

File: src/sagas/nanoContract.js [hunk 6]
```diff
@@ -167,31 +169,55 @@ export function* registerNanoContract({ payload }) {
...
+  const historyNewestToOldest = new Array(rawHistory.length)
+  for (let idx = 0; idx < rawHistory.length; idx += 1) {
+    const rawTx = rawHistory[idx];
     const network = wallet.getNetworkObject();
     const caller = addressUtils.getAddressFromPubkey(rawTx.nc_pubkey, network).base58;
     // XXX: Wallet Service Wallet doesn't implement isAddressMine.
     // It means this method can't run under wallet-service without
     // throwing an exception.
+    // XXX: Make sure to disable the Wallet Service feature if you are testing
+    // the nano-testnet network
```
Comment (andreabadesso) [inline src/sagas/nanoContract.js:220]: same here

File: src/sagas/nanoContract.js [hunk 7]
```diff
@@ -263,24 +284,38 @@ export function* requestHistoryNanoContract({ payload }) {
...
+      count: NANO_CONTRACT_TX_HISTORY_SIZE,
+    };
+    const { history } = yield call(fetchHistory, req);
 
-    if (after != null) {
-      // The first load has always `after` equals null. The first load means to be fast,
-      // but the subsequent ones are all request by user and we want slow down multiple
-      // calls to this effect.
+    if (after != null || before != null) {
+      // We want slow down multiple calls to this effect.
```
Comment (andreabadesso) [inline src/sagas/nanoContract.js:304]: Why? If you dispatch multiple actions at the same time, they will still execute at the same time, just 1000ms later

File: src/components/NanoContract/NanoContractDetails.js [hunk 8]
```diff
@@ -133,7 +136,7 @@ export const NanoContractDetails = ({ nc }) => {
             )}
             keyExtractor={(item) => item.txId}
             refreshing={isLoading}
-            onRefresh={handleMoreTransactions}
+            onRefresh={handleNewTransactions} // pull gesture
```
Comment (andreabadesso) [inline src/components/NanoContract/NanoContractDetails.js:139]: Ok, so this is the pull to refresh, but where is the code for the infinite scroll?

**Review by andreabadesso — CHANGES_REQUESTED** [review]

File: src/sagas/nanoContract.js (hunk 7 above)
Comment (alexruzenhack) [reply src/sagas/nanoContract.js:304]: You right! I removed it.

File: src/sagas/nanoContract.js (hunk 5 above)
Comment (alexruzenhack) [reply src/sagas/nanoContract.js:98]: I've opened an issue, see https://github.com/HathorNetwork/hathor-wallet-lib/issues/732.

File: src/components/NanoContract/NanoContractDetails.js (hunk 8 above)
Comment (alexruzenhack) [reply src/components/NanoContract/NanoContractDetails.js]: Sorry, I forgot to upload. Here it is: https://github.com/HathorNetwork/hathor-wallet-mobile/pull/512/commits/2040049f05e0125695c2ad5e612d54039da8b087

**Comment by alexruzenhack** [comment]
>I missed this in the add component PR, but can you rename this component? It doesn't make sense to have a NewHathorButton, how is it different from the HathorButton? Can we merge both?
-- Comment by @andreabadesso on https://github.com/HathorNetwork/hathor-wallet-mobile/pull/515, which is now incorporated here.

I will open a KTLO for this.

File: src/sagas/nanoContract.js [hunk 9]
```diff
@@ -167,33 +174,60 @@ export function* registerNanoContract({ payload }) {
...
-    // XXX: Wallet Service Wallet doesn't implement isAddressMine.
-    // It means this method can't run under wallet-service without
-    // throwing an exception.
-    // eslint-disable-next-line no-await-in-loop
-    const isMine = await wallet.isAddressMine(caller);
+    // XXX: Wallet Service doesn't implement isAddressMine.
+    // See issue: https://github.com/HathorNetwork/hathor-wallet-lib/issues/732
+    // Default to `false` if using Wallet Service.
+    let isMine = false;
+    const useWalletService = consumeGenerator(isWalletServiceEnabled());
```
Comment (andreabadesso) [inline src/sagas/nanoContract.js:225]: Why do you need that?

File: src/utils.js [hunk 10]
```diff
@@ -415,6 +415,18 @@ export const getNanoContractFeatureToggle = (state) => (
  */
 export const getTimestampFormat = (timestamp) => moment.unix(timestamp).format(t`DD MMM YYYY [•] HH:mm`)
 
+/**
+ * Extract the result of a generator function when it is done.
+ */
+export const consumeGenerator = (generator) => {
```
Comment (andreabadesso) [inline src/utils.js:421]: This is not "consuming" a generator, it's waiting until it's done and returning the last yielded value

Name is misleading

File: src/sagas/nanoContract.js (hunk 9 above)
Comment (andreabadesso) [reply src/sagas/nanoContract.js:225]: Can't you just use

`const useWalletService = yield call(isWalletServiceEnabled);`

?

**Review by andreabadesso — CHANGES_REQUESTED** [review]

File: src/utils.js (hunk 10 above)
Comment (alexruzenhack) [reply src/utils.js:421]: No, it is not returning the last yielded value, it is returning the return value of a generator after it has yield for the last time.

I believe it is "consuming" the generator because it is running the generator until its end.

File: src/sagas/nanoContract.js (hunk 9 above)
Comment (alexruzenhack) [reply src/sagas/nanoContract.js:225]: The function isn't a generator, but a promise. As we can't have async generators the only option to call one until its end is by consuming it. But in this case I'm interested in the result of it, not the yield values.

File: src/components/NanoContract/NanoContractDetails.js [hunk 11]
```diff
@@ -133,13 +137,53 @@ export const NanoContractDetails = ({ nc }) => {
...
+            ListFooterComponent={<LoadMoreButton lastTx={txHistory.slice(-1,).pop()} />}
+            extraData={[isLoading, error]}
           />
         )}
     </Wrapper>
   );
 };
 
+/**
+ * It show a button to 'Load More' transactions after the last one.
```
Comment (pedroferreira1) [inline src/components/NanoContract/NanoContractDetails.js:153]:
```suggestion
 * It shows a button to 'Load More' transactions after the last one.
```

File: src/components/NanoContract/NanoContractDetails.js [hunk 12]
```diff
@@ -133,13 +137,53 @@ export const NanoContractDetails = ({ nc }) => {
...
+ *
+ * @param {Object} prop Properties object
+ * @param {{
+ *   ncId: string;
+ *   txId: string;
+ * }} prop.lastTx A transaction item from transaction history
+ */
+const LoadMoreButton = ({ lastTx }) => {
+  const dispatch = useDispatch();
+  const isInitializeTx = lastTx.ncMethod === 'initialize';
```
Comment (pedroferreira1) [inline src/components/NanoContract/NanoContractDetails.js:164]: You can continue using this to check if it already loaded the last tx, but now the API returns a new value in the response `has_more`

**Review by pedroferreira1 — APPROVED** [review]

File: src/components/NanoContract/NanoContractDetails.js (hunk 12 above)
Comment (alexruzenhack) [reply src/components/NanoContract/NanoContractDetails.js:164]: Got it. Well, I prefer the current way. But if you want, I can open a KTLO to adapt the pagination to use the `has_more` property.
